### PR TITLE
feat(Datagrid): Batch actions

### PR DESCRIPTION
**Pull request by faraznaseem019**
Contributes to #1985

{{short description}}
I added an **Overflow Menu** to store all the buttons that are present in the default **Batch Actions** Datagrid table, and I added a **Select All** Button in the Datagrid Actions Toolbar, for when the user decides to click on a row. In the Overflow Menu, there are 4 buttons, each of which correspond to: Left Panel, Download CSV, Refresh Columns, and Row Size.

#### What did you change?
I added two additional functions to Datagrid.stories.js. These functions are called in the BatchActions function. 

- DatagridActionsBatchActions: Creates an overflow menu on the very left side which contains 4 buttons, and it renders a search bar next to the overflow menu. 
Screenshot 2022-06-07 at 4 48 59 PM

- BatchActionsDatagridBatchActions: Creates a modified version of the Datagrid Actions Toolbar, by adding a Select All Button that is adjacent (on the left side) to the Action button.
Screenshot 2022-06-07 at 4 49 08 PM

#### How did you test and verify your work?
I tested and verified my work, by manually testing the components that I created, and obtaining the result that I wanted.
For example, when the Row Size Dropdown menu is clicked on, it renders a perfectly visible dropdown menu of all the different sizes that are available.
Screenshot 2022-06-07 at 4 51 20 PM

In addition, when the select all rows radio button is clicked, I received a Datagrid Actions Toolbar with the Select All button present. Upon clicking on the button, I received an alert message with the text inside the alert window saying, 'Select All'.
Screenshot 2022-06-07 at 4 53 27 PM

### CODE REVIEW

**Review by holmansze — CHANGES_REQUESTED** [review]
@faraznaseem019 - I took a quick look and here're my initial comments:
- I tried the deploy preview but wasn't able to see the changes you made. I still see the dropdown for the header column checkbox, and I don't see the "select all" button in the toolbar. I tried both the "batch actions" and "select items in all pages" stories.

image


- Looks like the changes are made in the storybook story. I'd expect changes to be made in the actual Datagrid component and then surfaced via the storybook.

**Review by markkulube — CHANGES_REQUESTED** [review]
@faraznaseem019 

I may be misunderstanding #1985 but the ask is not to update the stories. 

It seems like a feature request so you programmatic changes need to be encapsulated within the `Datagrid` components. 

You probably can begin by porting the logic you've written in the story to the pre-existing `Datagrid` component i.e https://github.com/carbon-design-system/ibm-cloud-cognitive/blob/main/packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js.

If `BatchActions` is not a default behaviour, then the consumer can enable it via props.

**Review by faraznaseem019 — COMMENTED** [review]
Migrated css styles from storybook styles to datagrid scss

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js
Comment (holmansze) [inline packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js:38]: Same as above.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js
Comment (holmansze) [inline packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js:88]: Same as above.

**Review by holmansze — CHANGES_REQUESTED** [review]
@faraznaseem019 - thanks for the change. A few commnents:
- See the inline comments regarding moving the alert statements to the Storybook.
- Can you clean up some commented code?
- The outline for the checkboxes is truncated. Can you fix that?

image

**Review by markkulube — COMMENTED** [review]
@faraznaseem019 

Maybe check in with @matthewgallo or @holmansze if the `Left Panel`, `Download CSV `and `Refresh Columns` buttons should be functional, i.e. effecting actions and or rendering DOM elements as one would expect.

**Comment by matthewgallo** [comment]
Hey @markkulube, I think since the functionality behind these buttons will be left up to the consumer of the component it is ok to leave the `alert`s for now. We can revisit it later on after more of the overall Datagrid functionality has been complete.

**Comment by markkulube** [comment]
Thanks  for the feedback @matthewgallo. 

> Hey @markkulube, I think since the functionality behind these buttons will be left up to the consumer of the component it is ok to leave the `alert`s for now.
Leaving it to the consumer is consistent with the spirit of flexibility of the `Datagrid`.

> We can revisit it later on after more of the overall Datagrid functionality has been complete.
Sure thing. I imagine since refresh and download are generic enough actions, during later discussions, it maybe worth considering offering default handlers.

**Review by markkulube — CHANGES_REQUESTED** [review]
@faraznaseem019 

There are a few occurrences of commented code that probably should refactored out. I a pointed some of them out, but the entire change-set should be parsed to clean out needless comments.

Here are some more detailed thoughts commenting conventions.
https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/2062#pullrequestreview-1022357309

cc: @holmansze @matthewgallo @elycheea

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js]: Removed.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js]: Done.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js]: Removed.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js]: Removed.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js]: Removed variables.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid.stories.js]: Removed unnecessary code.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js:33]: Moved to Datagrid.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js:41]: Moved to Datagrid.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js:38]: Moved to Datagrid.

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js
Comment (faraznaseem019) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js:87]: I have kept this in DatagridToolbar.js, because this is one of the new functionalities that we are implementing in Batch Actions.

**Review by holmansze — COMMENTED** [review]
@faraznaseem019 @elycheea - in the design (issue 1985), it says:
```
Add More button when more than 4 actions present in the toolbar
```

Could you help me understand if this refers to:
- the blue toolbar when selections are made, or
- the general toolbar that is always there?

In the implementation, I saw an overflow menu in the general toolbar (that shows without selections being made). I would like to clarify if this is the intended behavior.

If yes, should it show up only when there are more than 4 actions present in the toolbar?

Also, the way the overflow menu is rendered seems a little different than what I usually see. It appears on the left (as opposed to right) and the settings button opens another submenu. I can't seem to see this in the design. Would you be able to clarify the intended behavior?

image

**Comment by elycheea** [comment]
@holmansze @faraznaseem019 The more button should be referring to the [blue batch actions toolbar](https://pages.github.ibm.com/cdai-design/pal/components/data-table/batchactions/usage#more-than-4-actions).

The DatagridActions that were initially in place should be correct — we can see them in the other stories as well. The row size menu should [always be aligned right](https://pages.github.ibm.com/cdai-design/pal/components/data-table/settings/usage#placement) as is documented.

It may be that we’ve misinterpreted when to use the OverflowMenu for the batch actions — I think this is specifically in place of the **More** menu in the blue batch actions toolbar.

**Comment by matthewgallo** [comment]
I've updated the BatchActions implementation so that it more closely matches the designs [here](https://pages.github.ibm.com/cdai-design/pal/components/data-table/batchactions/usage/). See interaction below:

https://user-images.githubusercontent.com/10215203/182938605-0b2d9584-5924-4e29-8fef-8d6374397c5b.mp4

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js
Comment (elycheea) [inline packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js:68]: From the guidelines, it looks like the more icon should use a chevron instead, which also means we may need it to be directional?

**Review by elycheea — COMMENTED** [review]
@matthewgallo Looks good but a few comments that I think we should review before this gets merged.
- More button should use chevrons instead of add based on the guideline?
- More button seems to use the width of the longest menu item — this may be a product of our button menu implementation, but it seems counterintuitive here since we drop to More to conserve space — so we should probably open an issue/enhancement for ButtonMenu to handle width.

We must have missed this in #2105 but it looks like the `--data-table--selected` class is getting used both for clickable row items and the selected state here. We may need a separate class for clickable rows. This can be a separate issue too unless you think it makes sense to address it here.

image

File: packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js
Comment (matthewgallo) [reply packages/cloud-cognitive/src/components/Datagrid/Datagrid/DatagridToolbar.js:68]: Unfortunately this (toggling icon based on open state) isn't possible with the `ButtonMenu` component _yet_ because of some underlying issues with the OverflowMenu from Carbon, but we do have an issue open https://github.com/carbon-design-system/ibm-cloud-cognitive/issues/2076 and also another with [Carbon](https://github.com/carbon-design-system/carbon/issues/11781). This is what @qbi11y has been working on

**Comment by matthewgallo** [comment]
Hey @elycheea, I've updated the `ButtonMenu` width issue and fixed the issue mentioned regarding the `--data-table--selected` class. The toggling of the chevron icons is blocked by the issues mentioned [here](https://github.com/carbon-design-system/ibm-cloud-cognitive/pull/2022#discussion_r943449404).

File: packages/cloud-cognitive/src/components/Datagrid/styles/_datagrid.scss
Comment (elycheea) [inline packages/cloud-cognitive/src/components/Datagrid/styles/_datagrid.scss:455]: Ahhh, good use of `:not` selector here to keep changes small. 👏

**Review by elycheea — APPROVED** [review]
Changes looking good! Thanks for documenting the ButtonMenu issue here so we hopefully will remember to update once we can.

**Pull request merged**
